### PR TITLE
Update to scintilla 5.5.2 & Lexilla 5.4.0

### PR DESCRIPTION
**Pull request by chcg**
Release 5.5.2 ( https://www.scintilla.org/scintilla552.zip )

    Released 21 August 2024.
    Add SCI_SETCOPYSEPARATOR for separator between parts of a multiple selection when copied to the clipboard. Feature #1530.
    Add SCI_GETUNDOSEQUENCE to determine whether an undo sequence is active and its nesting depth.
    Add SCI_STYLESETSTRETCH to support condensed and expanded text styles.
    Add SCI_LINEINDENT and SCI_LINEDEDENT. Feature #1524.
    Fix bug on Cocoa where double-click stopped working when system had been running for a long time.
    On Cocoa implement more values of font weight and stretch.

Release 5.4.0 ( https://www.scintilla.org/lexilla540.zip )

    Released 21 August 2024.
    Inside Lexilla, LexerModule instances are now const. This will require changes to applications that modify Lexilla.cxx, which may be done to add custom lexers.
    Lexer added for TOML "toml".
    Bash: Handle backslash in heredoc delimiter. Issue #257.
    Progress: Fix lexing of nested comments. Pull request #258.
    Force lower-casing of case-insensitive keyword lists so keywords match in some lexers. Issue #259.

### CODE REVIEW

File: scintilla/src/Editor.cxx [hunk 1]
```diff
@@ -4361,14 +4363,16 @@ void Editor::CopySelectionRange(SelectionText *ss, bool allowLineCopy) {
...
+		const std::string_view separator = (sel.selType == Selection::SelTypes::rectangle) ? pdoc->EOLString() : copySeparator;
+		for (size_t part = 0; part < rangesInOrder.size(); part++) {
+			text.append(RangeText(rangesInOrder[part].Start().Position(), rangesInOrder[part].End().Position()));
+			if ((sel.selType == Selection::SelTypes::rectangle) || (part < rangesInOrder.size() - 1)) {
+				// Append unless simple selection or last part of multiple selection
+				text.append(separator);
 			}
 		}
 		ss->Copy(text, pdoc->dbcsCodePage,
-			vs.styles[StyleDefault].characterSet, rangesInOrder.size() > 1, sel.selType == Selection::SelTypes::lines);
```
Comment (chcg) [inline scintilla/src/Editor.cxx:4371]: @donho Conflict between 4ff9d77d2bfbb323e2fcba10f02570710872db94 and the current modification by scintilla 5.5.2 update which still needs to be checked and adapted.

Comment (donho) [reply scintilla/src/Editor.cxx:4371]: @chcg 
IMO, for not having regression, `vs.styles[StyleDefault].characterSet, sel.IsRectangular(), sel.selType == Selection::SelTypes::lines);` should be `vs.styles[StyleDefault].characterSet, rangesInOrder.size() > 1, sel.selType == Selection::SelTypes::lines);`.
I'll test it and let you know the test result.

**Comment by ozone10** [comment]
follow up PR to enable `TOML` would be great
ref #5631

File: scintilla/src/Editor.cxx (hunk 1 above)
Comment (donho) [reply scintilla/src/Editor.cxx:4371]: @chcg 
After checking the old behaviour, the above modification is necessary, but not enough.
We should also change from:
`if ((sel.selType == Selection::SelTypes::rectangle) || (part < rangesInOrder.size() - 1)) {`
to
`if ((rangesInOrder.size() > 1) || (part < rangesInOrder.size() - 1)) {`

and as well from:
`const std::string_view separator = (sel.selType == Selection::SelTypes::rectangle) ? pdoc->EOLString() : copySeparator;`
to
`const std::string_view separator = (rangesInOrder.size() > 1) ? pdoc->EOLString() : copySeparator;`

**Review by donho — CHANGES_REQUESTED** [review]
Please change from:
`if ((sel.selType == Selection::SelTypes::rectangle) || (part < rangesInOrder.size() - 1)) {`
to
`if ((rangesInOrder.size() > 1) || (part < rangesInOrder.size() - 1)) {`

and also from:
`const std::string_view separator = (sel.selType == Selection::SelTypes::rectangle) ? pdoc->EOLString() : copySeparator;`
to
`const std::string_view separator = (rangesInOrder.size() > 1) ? pdoc->EOLString() : copySeparator;`

Finally change from
`vs.styles[StyleDefault].characterSet, sel.IsRectangular(), sel.selType == Selection::SelTypes::lines);` 
should be 
`vs.styles[StyleDefault].characterSet, rangesInOrder.size() > 1, sel.selType == Selection::SelTypes::lines);`.

in order to keep the same behaviour of Copy/Paste Multi-selection as previous version.

**Review by donho — APPROVED** [review]

**Pull request closed without merging**